File: src/main/java/io/myosotisdev/utopianism/util/SingleSelectionModel.java
package io.myosotisdev.utopianism.util;

import java.util.Collection;
import java.util.HashMap;
import java.util.Set;

public class SingleSelectionModel<K, V>
{
    private HashMap<K, V> map;
    private V current;

    public SingleSelectionModel()
    {
        this.map = new HashMap<>();
    }

    public HashMap<K, V> getModels()
    {
        return this.map;
    }

    public void add(K key, V value)
    {
        this.map.put(key, value);
    }

    public void addIfAbsent(K key, V value)
    {
        this.map.putIfAbsent(key, value);
    }

    public V get(K key)
    {
        return this.map.get(key);
    }

    public void setCurrentKey(K key)
    {
        setCurrent(this.map.get(key));
    }

    public void setCurrent(V val)
    {
        if(val != null)
            this.current = val;
    }

    public boolean containsKey(K key)
    {
        return map.containsKey(key);
    }

    public boolean containsValue(V value)
    {
        return map.containsValue(value);
    }

    public V current()
    {
        return this.current;
    }

    public Set<K> keys()
    {
        return map.keySet();
    }

    public Collection<V> values()
    {
        return map.values();
    }
}

File: src/main/java/io/myosotisdev/utopianism/util/SingleSelectionModel.kt
package io.myosotisdev.utopianism.util

class SingleSelectionModel<K, V>
{
    val models: HashMap<K, V> = HashMap()
    private var _hidden: K? = null
    var currentKey: K?
        get() = _hidden
        set(value)
        {
            _hidden = value
            current = models[value]
        }
    var current: V? = null

    operator fun set(key: K, value: V)
    {
        models[key] = value
    }

    fun addIfAbsent(key: K, value: V)
    {
        models.putIfAbsent(key, value)
    }

    operator fun get(key: K): V?
    {
        return models[key]
    }

    fun containsKey(key: K): Boolean
    {
        return models.containsKey(key)
    }

    fun containsValue(value: V): Boolean
    {
        return models.containsValue(value)
    }

    fun keys(): Set<K>
    {
        return models.keys
    }

    fun values(): Collection<V>
    {
        return models.values
    }
}